src/style/fog.js

```diff
@@ -35,83 +34,23 @@ const properties: Properties<Props> = new Properties({
-        const range = props.get('range');
-        const fogOpacity = smoothstep(FOG_PITCH_START, FOG_PITCH_END, pitch);
-        const fogStrength = props.get('strength');
-        const [start, end] = range;
-
-        // The fog is not physically accurate, so we seek an expression which satisfies a
-        // couple basic constraints:
-        //   - opacity should be 0 at the near limit
-        //   - opacity should be 1 at the far limit
-        //   - the onset should have smooth derivatives to avoid a sharp band
-        // To this end, we use an (1 - e^x)^n, where n is set to 3 to ensure the
-        // function is C2 continuous at the onset. The fog is about 99% opaque at
-        // the far limit, so we simply scale it and clip to achieve 100% opacity.
-        // https://www.desmos.com/calculator/3taufutxid
-        // The output of this function should match src/shaders/_prelude_fog.fragment.glsl
-        const decay = 6;
-        const t = (depthPx / heightPx - start) / (end - start);
-        let falloff = 1.0 - Math.min(1, Math.exp(-decay * t));
-
-        // Cube without pow()
-        falloff *= falloff * falloff;
-
-        // Scale and clip to 1 at the far limit
-        falloff = Math.min(1.0, 1.00747 * falloff);
-
-        // From src/render/painter.js via fog uniforms:
-        const fogExponent = 12 * Math.pow(1 - fogStrength, 2);
-
-        // Account for fog strength
-        falloff *= Math.pow(smoothstep(0, 1, t), fogExponent);
-
-        return falloff * fogOpacity;
-    }
-
-    getOpacityAtTileCoord(x: number, y: number, z: number, tileId: UnwrappedTileID, transform: Transform): number {
-        const mat = transform.calculateCameraMatrix(tileId);
-        const pos = [x, y, z];
-        vec3.transformMat4(pos, pos, mat);
-        const depth = vec3.length(pos);
-
-        return this.getFogOpacity(depth, transform.pitch, transform.height);
-    }
-
-    getFogOpacityAtLatLng(lngLat: LngLat, transform: Transform): number {
-        const meters = MercatorCoordinate.fromLngLat(lngLat);
-        const elevation = transform.elevation ? transform.elevation.getAtPoint(meters) : 0;
-        const pos = [meters.x, meters.y, elevation];
-        vec3.transformMat4(pos, pos, transform.mercatorFogMatrix);
-        const depth = vec3.length(pos);
-
-        return this.getFogOpacity(depth, transform.pitch, transform.height);
-    }
-}
-
 class Fog extends Evented {
     _transitionable: Transitionable<Props>;
     _transitioning: Transitioning<Props>;
     properties: PossiblyEvaluated<Props>;
-    sampler: FogSampler;
 
     constructor(fogOptions?: FogSpecification) {
         super();
         this._transitionable = new Transitionable(properties);
         this.set(fogOptions);
         this._transitioning = this._transitionable.untransitioned();
-        this.sampler = new FogSampler();
+    }
+
+    get state(): FogState {
+        return {
+            range: this.properties.get('range'),
```

Comment:
This throws an error `TypeError: Cannot read property 'get' of undefined` for me. It seems properties may or may not be exist at the time of this call.

Reply:
This has been fixed by deferring the marker update until the style has recalculated.